File: java/subsets_2.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List <List<Integer>> subsets = new ArrayList<>();
        subsets.add(new ArrayList<Integer>());

        int subsetSize = 0;

        for (int i = 0; i < nums.length; i++) {
            int startInd = (i >= 1 && nums[i] == nums[i - 1]) ? subsetSize : 0; // if duplicate then select from previous size list else add element to all the list
          
            subsetSize = subsets.size();
            for (int j = startInd; j < subsetSize; j++) {
                List <Integer> currentSubset = new ArrayList<>(subsets.get(j));
                currentSubset.add(nums[i]);
                subsets.add(currentSubset);
            }
        }
        return subsets;
    }

}

File: js/subsets_2.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 var subsetsWithDup = function(nums) {
    nums.sort();
    let subsets = [[]];
    let subsetLength = 0;
    
    for (let i = 0; i < nums.length; i++) {
        let startInd = (i >= 1 && nums[i] == nums[i-1]) ? subsetLength : 0; // if duplicate then select from previous size list else add element to all the list
        
        subsetLength = subsets.length;
        for (let j = startInd; j < subsetLength; j++) {
            let currSubset = [];
            if (subsets[j].length == 0) {
                currSubset = [nums[i]];
            } else {
                currSubset = [...subsets[j]];
                currSubset.push(nums[i]);
            }
            subsets.push(currSubset);
        }
    }
    
    return subsets;
};